node/src/accountant/mod.rs

```diff
@@ -745,133 +742,103 @@ impl Accountant {
         }
     }
 
-    fn handle_report_routing_service_provided_message(
-        &mut self,
-        msg: ReportRoutingServiceProvidedMessage,
-    ) {
-        let msg_id = self.msg_id();
-        trace!(
-            self.logger,
-            "Msg {}: Charging routing of {} bytes to wallet {}",
-            msg_id,
-            msg.payload_size,
-            msg.paying_wallet
-        );
-
-        let sum = self.record_service_provided(
-            msg.service_rate,
-            msg.byte_rate,
-            msg.timestamp,
-            msg.payload_size,
-            &msg.paying_wallet,
-        );
-
-        self.logging_utils.accounting_msgs_stats.manage_debug_log(
-            &self.logger,
-            AccountingMsgType::RoutingServiceProvided,
-            DEFAULT_ACCOUNTING_MSG_LOG_WINDOW,
-            vec![NewPosting::new(msg.paying_wallet.address(), sum)],
-        )
-    }
+    fn handle_report_service_provided_message<AccountingMessage>(&mut self, msg: AccountingMessage)
+    where
+        AccountingMessage: MessageWithServicesProvided,
```

Comment:
This should go away once you decide to use the structure instead of the trait implementation. The enum based approach will make things easier for you here. You can thank me later.